// eas-cli/build/graphql/queries/UpdateQuery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateQuery = exports.viewBranchUpdatesQueryUpdateLimit = void 0;
const tslib_1 = require("tslib");
const graphql_tag_1 = tslib_1.__importDefault(require("graphql-tag"));
const client_1 = require("../client");
exports.viewBranchUpdatesQueryUpdateLimit = 300;
exports.UpdateQuery = {
    async viewAllAsync({ appId }) {
        return (0, client_1.withErrorHandlingAsync)(client_1.graphqlClient
            .query((0, graphql_tag_1.default) `
            query ViewAllUpdates($appId: String!, $limit: Int!) {
              app {
                byId(appId: $appId) {
                  id
                  updateBranches(offset: 0, limit: $limit) {
                    id
                    name
                    updates(offset: 0, limit: $limit) {
                      id
                      group
                      message
                      createdAt
                      runtimeVersion
                      platform
                      actor {
                        id
                        ... on User {
                          username
                        }
                        ... on Robot {
                          firstName
                        }
                      }
                    }
                  }
                }
              }
            }
          `, {
            appId,
            limit: exports.viewBranchUpdatesQueryUpdateLimit,
        }, { additionalTypenames: ['UpdateBranch', 'Update'] })
            .toPromise());
    },
    async viewBranchAsync({ appId, name, limit = exports.viewBranchUpdatesQueryUpdateLimit, offset = 0, }) {
        return (0, client_1.withErrorHandlingAsync)(client_1.graphqlClient
            .query((0, graphql_tag_1.default) `
            query ViewBranchUpdates($appId: String!, $name: String!, $limit: Int!, $offset: Int!) {
              app {
                byId(appId: $appId) {
                  id
                  updateBranchByName(name: $name) {
                    id
                    name
                    updates(offset: $offset, limit: $limit) {
                      id
                      group
                      message
                      createdAt
                      runtimeVersion
                      platform
                      manifestFragment
                      actor {
                        id
                        ... on User {
                          username
                        }
                        ... on Robot {
                          firstName
                        }
                      }
                    }
                  }
                }
              }
            }
          `, {
            appId,
            name,
            limit,
            offset,
        }, { additionalTypenames: ['UpdateBranch', 'Update'] })
            .toPromise());
    },
};
